### ISSUE TITLE
Contents should be opaque to the did store and connection store

### ISSUE THREAD

**Issue by llorllale** [hyperledger/aries-framework-go#26]
I just realized after we closed #22: the store implementation will be able to read all the DIDs, including the pairwise mappings. This could result in a breach of privacy in the case where the storage is an external service.

Ideally, no stores should know what DIDs I am interacting with. This could be uniformly enforced across all implementations by a change in API. For example, a K-V store:

```go
// The layer that consumes this interface is responsible to 
// encrypting the data such that providers cannot interpret it.
// That layer remains secured in aries-framework-go.
type Store interface {
    Put(k string, v []byte)  error
    Get(k string) ([]byte, error)
}
```

See [RFC 0050 - Wallets](https://github.com/hyperledger/aries-rfcs/blob/master/concepts/0050-wallets/README.md#composition)

@rolsonquadras @troyronda thoughts?

**Comment by llorllale** [comment]
The main idea is that the Store should not be able to read what the Agent is persisting in it. This means the DID documents need to be encrypted **and then** given to the Store for persistence.

**We search by DID**
* A DID document's key is simply its DID. To prevent the Store from even reading the DID, we should pass in a hash of the DID as key
* The `kid` in incoming [encryption envelopes](https://github.com/hyperledger/aries-rfcs/tree/master/features/0019-encryption-envelope) is a [DID fragment](https://w3c-ccg.github.io/did-spec/#fragment) - searching for the document with that key should be straightforward (following from previous point: simply extract the DID from the `kid`, hash it, and use it as search parameter).

**Comment by llorllale** [comment]
The above is the bare minimum for the did-exchange demo. I suggest we start here and figure out details (like adding tags to the search capabilities) as we move forward.

**Comment by rolsonquadras** [comment]
@llorllale The above interface satisfies the document store defined in https://github.com/hyperledger/aries-framework-go/tree/master/pkg/storage/docstore. 

But we need to come up generic search function, particularly for https://github.com/hyperledger/aries-framework-go/blob/master/pkg/storage/connection/connstore.go#L41. By encrypting the value, there is no way to search based on the content. May be as you mentioned, we can defer the implementation to later.

**Comment by llorllale** [comment]
@rolsonquadras this method...

```go
LookupByTheirPublicKey(pubKey []byte) (*Connection, bool)
```

... can be replaced by using `Get(k string) ([]byte, error)` defined above in `Store`.

The way it would work is as follows:

1. The Agent will have previously saved the encrypted DID document with a hash of the DID using `Put(hash(did), encrypted(did_document))`
2. When receiving and unpacking a message, the Sender's `kid` will be a DID fragment, eg. `did:example:123456abcdefg#key1`
3. The Agent extracts `did:example:123456abcdefg` from the `kid`, hashes it, and looks up the sender's encrypted DID document from the Store using `Get(k string) ([]byte, error)`
4. The Agent decrypts the DID document and gets the key from it

**Comment by llorllale** [comment]
@rolsonquadras the above applies to receiving messages.

When sending a message, an Agent needs not only look up the recipient's DID document, but also  look up its own DID previously shared with the recipient during did-exchange.

So we're missing a step here.

Maybe the Agent can save this mapping to the store:
* `k`: hash("pair-did:example:123456abcdefg")
* `v`: encrypted({"mine": "did:example:56789ghjkl", "theirs", "did:example:123456abcdefg"})

In the demo I imagine this will look something along the lines of "Alice, send a msg to did:example:123456abcdefg", instead of using a human-readable identifier, ie. "Alice, send a msg to did:example:123456abcdefg".

**Comment by llorllale** [comment]
@rolsonquadras the above would involve three round trips to the Store when sending a message:

1. Lookup the mapping
2. Lookup my DID document
3. Lookup their DID document

The advantage of this approach is that the Store implementation is made very simple.

On the other hand, to minimize roundtrips we would need to push a little more logic into the Store. Perhaps we should simply adopt [Tags and Queries](https://github.com/hyperledger/aries-rfcs/blob/master/concepts/0050-wallets/README.md#tags-and-queries), along with a key.

**Comment by llorllale** [comment]
PR #31 merged, tags and queries will be tracked in #34
